ShareIt v2.3 add-bookings

Zazergel opened the pull request:
feat: Добавлена функциональность добавления комментариев, бронирования вещей.
feat: Данные теперь хранятся в БД, а не в памяти.

Zazergel committed:
ShareIt v2.3 add-bookings
feat: Добавлена функциональность добавления комментариев, бронирования вещей.
feat: Данные теперь хранятся в БД, а не в памяти.

Zazergel committed:
ShareIt v2.3.1 add-bookings
Попытка исправить ошибку из target #1

Zazergel committed:
ShareIt v2.3.2 add-bookings
Попытка исправить ошибку из target #2

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,21 +1,59 @@
+    @Column(name = "END_DATE", nullable = false)
+    LocalDateTime end;
+
+    @OneToOne
```
@ManyToOne Одна вещь может быть связана с несколькими бронированиями

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,21 +1,59 @@
+    @OnDelete(action = OnDeleteAction.CASCADE)
+    Item item;
+
+    @OneToOne
```
@ManyToOne Одна юзер может быть связан с несколькими бронированиями

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,49 @@
+
+    @GetMapping
+    public List<BookingResponseDto> getAllByBookerId(@RequestHeader(headerUserId) Long userId,
+                                                     @RequestParam(defaultValue = "ALL", required = false) String state) {
```
При указании дефолтного значения required = false не нужен

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,49 @@
+
+    @GetMapping("/owner")
+    public List<BookingResponseDto> getAllByOwnerId(@RequestHeader(headerUserId) Long userId,
+                                                    @RequestParam(defaultValue = "ALL", required = false) String state) {
```
При указании дефолтного значения required = false не нужен

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+        if (booking.getEnd().isEqual(booking.getStart())) {
+            throw new BookingException("Недопустимое время брони.");
+        }
+
```
Еще лучше проверить, что вещь не забронирована никем другим на запрашиваемые даты

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+        if (!repoBooking.getStatus().equals(Status.WAITING)) {
+            throw new BookingException("Ответ по бронированию уже дан.");
+        }
+
```
Лучше проверить, что бронирование еще началось по датам

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Comment.java:
```diff
@@ -0,0 +1,41 @@
+    @Column(name = "CREATED_DATE", nullable = false)
+    LocalDateTime created;
+
+    @OneToOne
```
@ManyToOne

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Comment.java:
```diff
@@ -0,0 +1,41 @@
+    @JoinColumn(name = "AUTHOR_ID", referencedColumnName = "ID", nullable = false)
+    User author;
+
+    @OneToOne
```
@ManyToOne

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -0,0 +1,53 @@
+    @Column(nullable = false)
+    Boolean available;
+
+    @OneToOne
```
@ManyToOne

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -0,0 +1,97 @@
+    public abstract Item toItem(ItemDto itemDto);
+
+    @Mapping(source = "owner", target = "ownerId", qualifiedByName = "mapUserIdFromUser")
+    @Mapping(target = "lastBooking", expression = "java(addLastBooking(item))")
```
Проблема N+1 запроса. Так как этот маппинг используется для списка вещей, то лучше вытащить из базы все бронирования для всего списка вещей одним запросом.
Аналогично для addNextBooking(...) и addComment(...)

Zazergel committed:
ShareIt v2.4.0 add-bookings
refactor: Рефакторинг мапперов ItemMapper, BookingMapper
del: Удалены классы MapperService и MapperServiceImpl
feat: Добавлен класс StateException
fix: Исправлены аннотации OneToOne на ManyToOne в классах: Comment,  Item, Booking
del: Удалены лишние поля в DTO классах: BookingRequestDto, CommentRequestDto

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,191 @@
+            throw new BookingException("Предмет недоступен для бронирования.");
+        }
+        if (!bookingRepository
+                .findByItemIdAndStartBeforeAndEndAfterAndStatusEqualsOrderByStartAsc(
```
Почему проверка только полностью вложенного интервала времени? А если новое и одно из существующих бронирований пересекаются лишь частично?

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -3,79 +3,169 @@
+        log.info("Вывод всех вещей пользователя с id {}.", userId);
+
+        return itemRepository.getAllByOwnerIdOrderByIdAsc(userId).stream()
+                .map((item) -> itemMapper.toItemExtendedDto(item, addLastBooking(item),
```
Здесь для каждого item-а вызывается метод addLastBooking(...), в котором выполняется запрос в БД. То есть если у нас N item-ов, то будет N дополнительных запросов в БД. А нужно вытащить из БД все бронирования одним запросом.
Аналогично для addNextBooking(...) и для addComment(...)

Zazergel committed:
ShareIt v2.5.0
fix: Исправлена проблема N+1 в методе getByOwnerId(Long userId) класса ItemServiceImpl
feat: добавлена проверка пересечений по датам для новых бронирований в методе add(Long userId) класса BookingServiceImpl

Zazergel committed:
ShareIt v2.5.1
fix: исправлены ошибки codeStyle

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -3,79 +3,228 @@
+                nextBookingByItem.put(item.getId(), bookingDtosByItem.get(item.getId())
+                        .stream()
+                        .filter(b -> b.getStart().isBefore(LocalDateTime.now()))
+                        .findFirst().orElse(null));
```
Точно ли findFirst() ?
Будет найден самый первый с датой до now(), но он не будет гарантированно самым ближайшим к now()

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,202 @@
+        if (!item.getAvailable()) {
+            throw new BookingException("Предмет недоступен для бронирования.");
+        }
+        if (!bookingRepository.findByItemIdAndStartBeforeAndEndAfterAndStatusEqualsOrderByStartAsc(item.getId(),
```
Тяжелая проверка с использованием 4 запросов в БД)
